### PR TITLE
Fix an issue that caused Lightning to extract the batch size even though it was set by the user in `LightningModule.log`

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Fixes: https://github.com/PyTorchLightning/pytorch-lightning/issues/10349#issuecomment-962490098
Fixes: #10576

Seems like after we added the warning to check for batch_size for an ambiguous batch, a few users realized that the batch_size used internally isn't correct. We always make a call to extract batch size even when it's set by the user explicitly inside `self.log(batch_size=...)` which IMO is not required because we have a recursive call to extract batch size which causes unnecessary extraction for it even when it's not required plus it will raise a warning if someone specifies it and we still look for it. So now the call will be made inside `.log` only when batch_size is not specified.

In short: this fixes an issue that caused Lightning to extract the batch size even though it was set by the user in `LightningModule.log`.

So an overview of the solution is:
track the batch and use it to extract batch_size inside `self.log` call when it's not specified explicitly and metric is being logged with `on_epoch=True` and will be averaged at the end.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @edward-io @ananthsub

### CODE REVIEW

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 1]
```diff
@@ -194,8 +192,6 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
             with self.trainer.profiler.profile("run_training_batch"):
                 batch_output = self.batch_loop.run(batch, batch_idx)
 
-        self.trainer._results.batch_size = batch_size
```
Comment (rohitgr7) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:197]: reviewer: this is not required since we don't keep track of it now. if it's needed, it will be extracted using the `current_batch` attached to `ResultCollection`

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 2]
```diff
@@ -253,28 +252,20 @@ def _log_gpus_metrics(self) -> None:
...
     def on_epoch_start(self) -> None:
         self._epoch_end_reached = False
 
-    def on_batch_start(self, batch_idx: int, batch: Any) -> int:
+    def on_batch_start(self, batch_idx: int, batch: Any) -> None:
         self._batch_idx = batch_idx
         self._epoch_end_reached = False
-        return self.on_new_batch(batch)
+        assert self.trainer._results is not None
+        self.trainer._results.current_batch = batch
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:262]: reviewer: we keep track of current batch instead of batch size since we are extracting it inside `self.log` call.

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -358,122 +359,25 @@ def log(
             rank_zero_only: Whether the value will be logged only on rank 0. This will prevent synchronization which
                 would produce a deadlock as not all processes would perform this log call.
         """
-        if tbptt_reduce_fx is not None:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/core/lightning.py]: Please, do not do this change here, keep the PR focused to the fix. This change should be done separately if so.

File: CHANGELOG.md [hunk 4]
```diff
@@ -116,16 +116,14 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
...
 
-- Fixed an issue where the model wrapper in Lite converted non-floating point tensors to float ([#10429](https://github.com/PyTorchLightning/pytorch-lightning/pull/10429))
+- Fixed dataloader workers with `persistent_workers` being deleted on every iteration ([#10434](https://github.com/PyTorchLightning/pytorch-lightning/pull/10434))
 
 
--
-
+- Fixed an issue where the model wrapper in Lite converted non-floating point tensors to float ([#10429](https://github.com/PyTorchLightning/pytorch-lightning/pull/10429))
 
--
```
Comment (carmocca) [inline CHANGELOG.md:128]: Keep these, they are here to try and avoid conflicts after the CHANGELOG is empty

**Review by carmocca — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:262]: Why not put that inside `on_new_batch`?

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 5]
```diff
@@ -356,7 +355,7 @@ class ResultCollection(dict):
     def __init__(self, training: bool, device: Optional[Union[str, torch.device]] = None) -> None:
         super().__init__()
         self.training = training
-        self._batch_size = torch.tensor(1, device=device)
+        self._current_batch = None
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:358]: I'm not a fan of this decision. Saving only the batch size value was done for memory and performance savings.

If each result object saves the current batch, then it's on more reference that we need to clean-up

Why can't the loop do it? I don't want the loop batch logic to leak into the ResultCollection

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:262]: if we do that and someone specifies `self.log(batch_size=...)` and uses an ambiguous batch, then it will raise a warning suggesting user to use `self.log(batch_size=...)` which seems odd. so we can't really extract the batch_size outside the `LightningModule.log` because we don't know until then whether the user specified the batch_size explicitly or not.

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 5 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:358]: this is cleaned inside `on_batch_end`. we somehow need to access the current batch inside `LightningModule.log`, can't find an alternative way for it. any suggestions?

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:262]: also in the current implementation if someone is not even logging, they will get this warning because we try to extract the batch size regardless of the fact whether its required or not

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -428,6 +429,29 @@ def log(
                 "With `def training_step(self, dataloader_iter)`, `self.log(..., batch_size=...)` should be provided."
             )
 
+        reduce_fx = reduce_fx.lower() if isinstance(reduce_fx, str) else reduce_fx
+
+        # check if we have extracted the batch size already
```
Comment (tchaton) [inline pytorch_lightning/core/lightning.py:434]: Can we move this logic within the ResultCollection instead?

Comment (tchaton) [reply pytorch_lightning/core/lightning.py:434]: I think the ResultCollection should keep track of the batch and recompute the batch_size only if the batch changed and log is been performed.

Comment (rohitgr7) [reply pytorch_lightning/core/lightning.py]: done! result collection keeps track of the batch, and update & stores the batch size whenever it's required.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:197]: It was added by @carmocca as a fix in #9791 (checked git blame) so I am rather confused by your comment 😅

File: CHANGELOG.md [hunk 7]
```diff
@@ -121,6 +121,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed `CombinedLoader` and `max_size_cycle` didn't receive a `DistributedSampler` ([#10374](https://github.com/PyTorchLightning/pytorch-lightning/issues/10374))
 
 
+- Fixed the call to `extract_batch_size` only when its required ([#10408](https://github.com/PyTorchLightning/pytorch-lightning/pull/10408))
```
Comment (awaelchli) [inline CHANGELOG.md:124]: The changelog is not clear what the issue was that got fixed. IMO it should always mention the problem. Describing the solution is optional.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 1 above)
Comment (rohitgr7) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:197]: it's changed a bit now.. we do keep track of it inside resultCollection. For tbptt I checked and found we don't split it across the batch_size dimension, rather we do it across time dimension

File: CHANGELOG.md (hunk 7 above)
Comment (rohitgr7) [reply CHANGELOG.md:124]:
```suggestion
- Fixed an issue to not extract batch_size when batch_size inside `LightningModule.log` is set by the user ([#10408](https://github.com/PyTorchLightning/pytorch-lightning/pull/10408))
```

Comment (rohitgr7) [reply CHANGELOG.md:124]: better?

Comment (awaelchli) [reply CHANGELOG.md:124]: Sorry, I think it is worse 😅 

I think it is either

> Fixed an issue that prevented Lightning to extract the batch size even though it was set by the user in LightningModule.log


or

> Fixed an issue that caused Lightning to extract the batch size even though it was set by the user in LightningModule.log


Totally not sure what you are trying to fix. Given the description of the PR, I probably think it is the latter.

Comment (rohitgr7) [inline CHANGELOG.md:124]:
```suggestion
- Fixed an issue that caused Lightning to extract the batch size even though it was set by the user in `LightningModule.log` ([#10408](https://github.com/PyTorchLightning/pytorch-lightning/pull/10408))
```

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:262]: I believe this line should clean the previous batch_size.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 8]
```diff
@@ -290,6 +282,9 @@ def on_batch_end(self) -> None:
         self._progress_bar_metrics.update(metrics["pbar"])
         self._callback_metrics.update(metrics["callback"])
         self._logged_metrics.update(metrics["log"])
+        assert self.trainer._results is not None
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:285]:
```suggestion
        # drop reference to current batch and batch_size
        assert self.trainer._results is not None
```

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 9]
```diff
@@ -253,28 +252,21 @@ def _log_gpus_metrics(self) -> None:
...
         self._epoch_end_reached = False
 
-    def on_batch_start(self, batch_idx: int, batch: Any) -> int:
+    def on_batch_start(self, batch_idx: int, batch: Any) -> None:
         self._batch_idx = batch_idx
         self._epoch_end_reached = False
-        return self.on_new_batch(batch)
+        assert self.trainer._results is not None
+        self.trainer._results.current_batch = batch
+        self.trainer._results.current_batch_size = None
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:263]: Add a comment why this line needs to be there.

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:380]: @carmocca We moved the logic within the ResultCollection as the other alternative was to keep this logic within the LightningModule.log method. This would make Trainer / LightningModule separation more complex.

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 10]
```diff
@@ -458,10 +467,27 @@ def log(
                 f"You called `self.log({name}, ...)` twice in `{fx}` with different arguments. This is not allowed"
             )
 
-        if batch_size is not None:
-            self.batch_size = batch_size
+        # check if we have extracted the batch size already
+        if batch_size is None:
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:471]: Let's move this logic within its own private method

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 11]
```diff
@@ -575,19 +601,10 @@ def fn(item: ResultMetric) -> None:
...
-        except RecursionError:
-            batch_size = 1
-        self.batch_size = batch_size  # the setter converts it to `Tensor`
-        return batch_size
-
     def to(self, *args: Any, **kwargs: Any) -> "ResultCollection":
         """Move all data to the given device."""
         self.update(apply_to_collection(dict(self), (torch.Tensor, Metric), move_data_to_device, *args, **kwargs))
 
-        self._batch_size = self._batch_size.to(*args, **kwargs)
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:590]: This would still be required with current_batch_size if we were to move the ResultCollection on CPU mid batch iteration.

File: tests/trainer/logging_/test_train_loop_logging.py [hunk 12]
```diff
@@ -715,19 +715,15 @@ def on_validation_epoch_end(self):
             assert all(v == 3 for v in self.trainer.callback_metrics.values())
 
         def on_train_batch_start(self, batch, batch_idx):
-            assert self.trainer._results.batch_size == 2
```
Comment (tchaton) [inline tests/trainer/logging_/test_train_loop_logging.py:718]: Why did you remove this ?

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 11 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:590]: I wonder do we even need to move the batch_size to the device at all. It's a scalar and wherever it's used, it's used along with other tensors which are already on their respective devices.

```bash
>>> a = torch.tensor(10.).cuda()
>>> a + 7
torch(17., device='cuda:0')
>>> a * 7
torch(70., device='cuda:0')
```

File: tests/trainer/logging_/test_train_loop_logging.py (hunk 12 above)
Comment (rohitgr7) [reply tests/trainer/logging_/test_train_loop_logging.py:718]: because reduce_fx = 'sum', so we don't track the batch_size in such a case.

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 8 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:285]: done

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py (hunk 9 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:263]: done

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 10 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:471]: done

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:590]: This is done to ensure there is no memory left on the GPU. We have tests for this and they would fail.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 13]
```diff
@@ -390,14 +392,22 @@ def append_fn(v: ResultMetric) -> None:
...
+    def _extract_batch_size(self, batch_size: Optional[int], meta: _Metadata) -> int:
+        # check if we have extracted the batch size already
+        batch_size = batch_size or self.batch_size
+        if batch_size is not None:
+            return batch_size
+        batch_size = 1
+
+        # extract it
+        fx_validate = _FxValidator.functions.get(meta.fx.split(".")[0])
+        if meta.on_epoch and fx_validate is not None and (True in fx_validate["on_step"]) and meta.is_mean_reduction:
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:404]: why exactly do we need the `True in fx_validate["on_step"]` check?

**Review by carmocca — COMMENTED** [review]
LGTM! I took the liberty of pushing several commits since it was easier than posting comments. Please take a look at the changes.

Two comments and you'll have my approval ✅

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 14]
```diff
@@ -390,14 +392,22 @@ def append_fn(v: ResultMetric) -> None:
...
+            try:
+                batch_size = extract_batch_size(self.batch)
+                self.batch_size = batch_size
+            except RecursionError:
```
Comment (ninginthecloud) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:408]: @rohitgr7  There could be other exceptions coming from `extract_batch_size()`. For example, if batch is ```
batch={'a':torch.tensor([1, 2, 3, 4, 5]), 'b': [torch.tensor(1), torch.tensor(2)]}
```
it raises index error when batch.size(0) is called for torch.tensor(1). 
I understand the good intention to automatically retrieve batch_size from batch, but we can't control how user set up their batch input. It's better to involve  user to provide batch_size.

File: tests/trainer/logging_/test_train_loop_logging.py [hunk 15]
```diff
@@ -749,3 +746,36 @@ def validation_epoch_end(self, *_) -> None:
...
+def test_no_batch_size_extraction_with_specifying_explictly(tmpdir):
+    batch_size = BoringModel().train_dataloader().batch_size + 1
+    fast_dev_run = 2
+    log_val = 7
+
+    class CustomBoringModel(BoringModel):
+        def on_before_batch_transfer(self, batch, *args, **kwargs):
+            # This is an ambiguous batch which have multiple potential batch sizes
+            if self.trainer.training:
+                batch = {"batch1": torch.randn(batch_size, 10), "batch2": batch}
```
Comment (ninginthecloud) [inline tests/trainer/logging_/test_train_loop_logging.py:760]: Hi, @rohitgr7  could we add more test cases to reflect users have mentioned in issue #10576 ?  This test case only captures a very ideal situation.

File: tests/trainer/logging_/test_train_loop_logging.py [hunk 16]
```diff
@@ -749,3 +746,36 @@ def validation_epoch_end(self, *_) -> None:
     train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
     val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
     trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
+
+
+def test_no_batch_size_extraction_with_specifying_explictly(tmpdir):
```
Comment (ninginthecloud) [inline tests/trainer/logging_/test_train_loop_logging.py:751]: Maybe I have missed here. do we have test case when user set batch_size in `LightningModule.log` and our code did not spam warning for ambiguous batch?

**Review by ninginthecloud — COMMENTED** [review]
Hi, @rohitgr7  Thanks a lot for tackling this issue! I left a few comments in code, feel free to let me know if you have any questions. 
A generic idea from my point of view, I think `extract_data_batch()` is this kind function we could convert as hook and delegate to user to implement, because they know their batch, it will be easily for them to define the batch_size or how to get batch_size. While here we had a lot assumptions about what the batch is going to look like, but we still can only extract valid batch_size under very limited situations.  
Anyway, I don't want to interrupt this PR discussion thread, I will initialize an issue to start the discussion. 
cc: @tchaton @ananthsub

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 17]
```diff
@@ -390,14 +392,22 @@ def append_fn(v: ResultMetric) -> None:
...
-    def batch_size(self) -> torch.Tensor:
-        # performance: cache the `batch_size` tensor instead of re-creating it
-        return self._batch_size
-
-    @batch_size.setter
-    def batch_size(self, value: int) -> None:
-        self._batch_size = torch.tensor(value, device=self.device)
+    def _extract_batch_size(self, batch_size: Optional[int], meta: _Metadata) -> int:
+        # check if we have extracted the batch size already
+        batch_size = batch_size or self.batch_size
```
Comment (rohitgr7) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:397]: @carmocca thanks for updates.. looks good.
this will fail with `self.log(..., batch_size=0)`. if the user is making a mistake, we shouldn't touch it.

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 13 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:404]: well, we need to ensure that batch_size should only be extracted for the cases where step-level logging is allowed. Also in other cases `results.batch` will be None too. We can avoid this check by allowing the call for `extract_batch_size(None)` and returning 1 from there, but it didn't seem ideal to me.

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 18]
```diff
@@ -390,14 +392,25 @@ def append_fn(v: ResultMetric) -> None:
...
+        if batch_size is not None:
+            return batch_size
 
-    @batch_size.setter
-    def batch_size(self, value: int) -> None:
-        self._batch_size = torch.tensor(value, device=self.device)
+        # extract it
+        fx_validate = _FxValidator.functions.get(meta.fx.split(".")[0])
```
[the remark this answers is not in the document]
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:403]: we append the dataloader_idx to meta.fx if we have multiple dataloaders

File: tests/trainer/logging_/test_train_loop_logging.py (hunk 16 above)
Comment (rohitgr7) [reply tests/trainer/logging_/test_train_loop_logging.py:751]: yes. 2nd log inside `training_step`
```
 self.log("epoch_log_val", log_val, batch_size=batch_size, ...)
```

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 14 above)
Comment (rohitgr7) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:408]: good catch! I guess this is the only edge case we had with tensors. Added a check for it.

**Comment by tchaton** [comment]
> Hi, @rohitgr7 Thanks a lot for tackling this issue! I left a few comments in code, feel free to let me know if you have any questions. A generic idea from my point of view, I think `extract_data_batch()` is this kind function we could convert as hook and delegate to user to implement, because they know their batch, it will be easily for them to define the batch_size or how to get batch_size. While here we had a lot assumptions about what the batch is going to look like, but we still can only extract valid batch_size under very limited situations. Anyway, I don't want to interrupt this PR discussion thread, I will initialize an issue to start the discussion. cc: @tchaton @ananthsub

Hey @ninginthecloud, I believe this is a great idea and this would make the logic more efficient. Personally, I am in favor of this.

@awaelchli @carmocca

**Comment by rohitgr7** [comment]
> Hi, @rohitgr7 Thanks a lot for tackling this issue! I left a few comments in code, feel free to let me know if you have any questions. A generic idea from my point of view, I think `extract_data_batch()` is this kind function we could convert as hook and delegate to user to implement, because they know their batch, it will be easily for them to define the batch_size or how to get batch_size. While here we had a lot assumptions about what the batch is going to look like, but we still can only extract valid batch_size under very limited situations. Anyway, I don't want to interrupt this PR discussion thread, I will initialize an issue to start the discussion. cc: @tchaton @ananthsub

if we do this, then we will have two different ways by which user can define their batch size. One will be this hook and another one will be `self.log(..., batch_size=)`. They still can define methods within their lightning module and can call it to retrieve the batch_size and use it within `self.log`. Plus we already have too many hooks and this seems more like a utility method to me. Just a personal opinion :)

**Comment by tchaton** [comment]
To add to @rohitgr7 point,

> self.log(..., batch_size=) was more a rescue solution and I personally don't find it elegant to provide the batch size, especially if you are logging outside of the *_step methods of the LightningModule where the current batch is unavailable. But this might affect a minority of users.

> Plus we already have too many hooks
This is entirely true and we want to enforce only 1 way to do anything within Lightning. I believe we would have to weigh the hook vs the self.log(..., batch_size=) in terms of usability/simplicity. 

@carmocca @awaelchli @SeanNaren  Any thoughts ?

**Comment by rohitgr7** [comment]
@tchaton let's discuss more here: https://github.com/PyTorchLightning/pytorch-lightning/issues/10635

File: tests/trainer/logging_/test_train_loop_logging.py (hunk 15 above)
Comment (rohitgr7) [reply tests/trainer/logging_/test_train_loop_logging.py:760]: well this test is just to check in what cases we should make a call to `extract_batch_size`. for the warning/error check we have tests here: https://github.com/PyTorchLightning/pytorch-lightning/blob/29797720cdbd74f5440c8e7186c82fe39b2002aa/tests/utilities/test_data.py#L18

Since I am not super familiar with torch-geometric, feel free to add some more here :)

**Review by carmocca — APPROVED** [review]

File: tests/trainer/logging_/test_train_loop_logging.py (hunk 15 above)
Comment (carmocca) [reply tests/trainer/logging_/test_train_loop_logging.py:760]: Resolving this, we can add more when we find new examples that do not work
